{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sragent\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `sragent` tool is centered on a single funciton `gather()`. \\\n",
    "The basic function of `gather()` is to take a Bio Project ID and access the Sequence Read Archive (SRA) to collect metadata on all of the samples within that project.\\\n",
    "It does this using the `entrez` tools from BioPython and some cluncky XML parsing to extract useful metadata.\\\n",
    "`gather()` will accept a single ID string or a list of strings for any number of projects.\\\n",
    ">[!WARNING]\n",
    ">`gather()` accesses an external server that is prone to connection problems.\\\n",
    "> You may encounter connection errors. Usually waiting and rerunning solves this, but you may need to split your request into chuncks.\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this demo project I'm going to try to pull and annotate a large number of Yeast Histone PTM ChIP-seq experiments. \\\n",
    "Below are a list of BioProjecIDs I've already collected that include any experiment profiling a Histone PTM. \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "projects = ['PRJNA262623', 'PRJNA227448', 'PRJNA140547', 'PRJNA989169',\n",
    "            'PRJNA954824', 'PRJNA912607', 'PRJNA831793', 'PRJNA783027',\n",
    "            'PRJNA753826', 'PRJNA737490', 'PRJNA721183', 'PRJNA672715',\n",
    "            'PRJNA643248', 'PRJNA588479', 'PRJNA559331', 'PRJNA492238',\n",
    "            'PRJNA487157', 'PRJNA450434', 'PRJNA384583', 'PRJNA320298',\n",
    "            'PRJNA278334', 'PRJNA274975', 'PRJNA254082', 'PRJNA231240',\n",
    "            'PRJNA153387']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now run `gather()` with this list of IDs. \\\n",
    "By default, we don't have to provide any other arguments to just get metadata for these projects. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching PRJNA262623...\n",
      "PRJNA262623 fetch complete...\n",
      "Fetching PRJNA227448...\n",
      "PRJNA227448 fetch complete...\n",
      "Fetching PRJNA140547...\n",
      "PRJNA140547 fetch complete...\n",
      "Fetching PRJNA989169...\n",
      "PRJNA989169 fetch complete...\n",
      "Fetching PRJNA954824...\n",
      "PRJNA954824 fetch complete...\n",
      "Fetching PRJNA912607...\n",
      "PRJNA912607 fetch complete...\n",
      "Fetching PRJNA831793...\n",
      "PRJNA831793 fetch complete...\n",
      "Fetching PRJNA783027...\n",
      "PRJNA783027 fetch complete...\n",
      "Fetching PRJNA753826...\n",
      "PRJNA753826 fetch complete...\n",
      "Fetching PRJNA737490...\n",
      "PRJNA737490 fetch complete...\n",
      "Fetching PRJNA721183...\n",
      "PRJNA721183 fetch complete...\n",
      "Fetching PRJNA672715...\n",
      "PRJNA672715 fetch complete...\n",
      "Fetching PRJNA643248...\n",
      "PRJNA643248 fetch complete...\n",
      "Fetching PRJNA588479...\n",
      "PRJNA588479 fetch complete...\n",
      "Fetching PRJNA559331...\n",
      "PRJNA559331 fetch complete...\n",
      "Fetching PRJNA492238...\n",
      "PRJNA492238 fetch complete...\n",
      "Fetching PRJNA487157...\n",
      "PRJNA487157 fetch complete...\n",
      "Fetching PRJNA450434...\n",
      "PRJNA450434 fetch complete...\n",
      "Fetching PRJNA384583...\n",
      "PRJNA384583 fetch complete...\n",
      "Fetching PRJNA320298...\n",
      "PRJNA320298 fetch complete...\n",
      "Fetching PRJNA278334...\n",
      "PRJNA278334 fetch complete...\n",
      "Fetching PRJNA274975...\n",
      "PRJNA274975 fetch complete...\n",
      "Fetching PRJNA254082...\n",
      "PRJNA254082 fetch complete...\n",
      "Fetching PRJNA231240...\n",
      "PRJNA231240 fetch complete...\n",
      "Fetching PRJNA153387...\n",
      "PRJNA153387 fetch complete...\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_id</th>\n",
       "      <th>project_title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>protocol</th>\n",
       "      <th>run_id</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>title</th>\n",
       "      <th>organism</th>\n",
       "      <th>assay_id</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593257</td>\n",
       "      <td>SRX717562</td>\n",
       "      <td>input2.2_15</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593256</td>\n",
       "      <td>SRX717561</td>\n",
       "      <td>input1.2_15</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593255</td>\n",
       "      <td>SRX717560</td>\n",
       "      <td>input2.2_8</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593254</td>\n",
       "      <td>SRX717559</td>\n",
       "      <td>input3.1_8</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593253</td>\n",
       "      <td>SRX717558</td>\n",
       "      <td>input1.3_60</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA153387</td>\n",
       "      <td>Co-dependency of H2B monoubiquitination and nu...</td>\n",
       "      <td>Monoubiquitination of histone H2B on lysine 12...</td>\n",
       "      <td>NO_PROTOCOL</td>\n",
       "      <td>SRR441515</td>\n",
       "      <td>SRX127920</td>\n",
       "      <td>delta_chd1_K79me3</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : delta_Chd1 yeast (FM391,delta Ch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA153387</td>\n",
       "      <td>Co-dependency of H2B monoubiquitination and nu...</td>\n",
       "      <td>Monoubiquitination of histone H2B on lysine 12...</td>\n",
       "      <td>NO_PROTOCOL</td>\n",
       "      <td>SRR441512</td>\n",
       "      <td>SRX127919</td>\n",
       "      <td>delta_chd1_wce</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : delta_Chd1 yeast (FM391,delta Ch...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA153387</td>\n",
       "      <td>Co-dependency of H2B monoubiquitination and nu...</td>\n",
       "      <td>Monoubiquitination of histone H2B on lysine 12...</td>\n",
       "      <td>NO_PROTOCOL</td>\n",
       "      <td>SRR441509</td>\n",
       "      <td>SRX127918</td>\n",
       "      <td>wt_FM391_K4me3</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : Wild-type yeast (FM391) genotype...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA153387</td>\n",
       "      <td>Co-dependency of H2B monoubiquitination and nu...</td>\n",
       "      <td>Monoubiquitination of histone H2B on lysine 12...</td>\n",
       "      <td>NO_PROTOCOL</td>\n",
       "      <td>SRR441506</td>\n",
       "      <td>SRX127917</td>\n",
       "      <td>wt_FM391_K79me3</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : Wild-type yeast (FM391) genotype...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PRJNA153387</td>\n",
       "      <td>Co-dependency of H2B monoubiquitination and nu...</td>\n",
       "      <td>Monoubiquitination of histone H2B on lysine 12...</td>\n",
       "      <td>NO_PROTOCOL</td>\n",
       "      <td>SRR441503</td>\n",
       "      <td>SRX127916</td>\n",
       "      <td>wt_FM391_wce</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : Wild-type yeast (FM391) genotype...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1172 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     project_id                                      project_title  \\\n",
       "0   PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "1   PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "2   PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "3   PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "4   PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "..          ...                                                ...   \n",
       "1   PRJNA153387  Co-dependency of H2B monoubiquitination and nu...   \n",
       "2   PRJNA153387  Co-dependency of H2B monoubiquitination and nu...   \n",
       "3   PRJNA153387  Co-dependency of H2B monoubiquitination and nu...   \n",
       "4   PRJNA153387  Co-dependency of H2B monoubiquitination and nu...   \n",
       "5   PRJNA153387  Co-dependency of H2B monoubiquitination and nu...   \n",
       "\n",
       "                                             abstract  \\\n",
       "0   Covalent histone modifications are highly cons...   \n",
       "1   Covalent histone modifications are highly cons...   \n",
       "2   Covalent histone modifications are highly cons...   \n",
       "3   Covalent histone modifications are highly cons...   \n",
       "4   Covalent histone modifications are highly cons...   \n",
       "..                                                ...   \n",
       "1   Monoubiquitination of histone H2B on lysine 12...   \n",
       "2   Monoubiquitination of histone H2B on lysine 12...   \n",
       "3   Monoubiquitination of histone H2B on lysine 12...   \n",
       "4   Monoubiquitination of histone H2B on lysine 12...   \n",
       "5   Monoubiquitination of histone H2B on lysine 12...   \n",
       "\n",
       "                                             protocol      run_id  \\\n",
       "0   Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593257   \n",
       "1   Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593256   \n",
       "2   Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593255   \n",
       "3   Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593254   \n",
       "4   Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593253   \n",
       "..                                                ...         ...   \n",
       "1                                         NO_PROTOCOL   SRR441515   \n",
       "2                                         NO_PROTOCOL   SRR441512   \n",
       "3                                         NO_PROTOCOL   SRR441509   \n",
       "4                                         NO_PROTOCOL   SRR441506   \n",
       "5                                         NO_PROTOCOL   SRR441503   \n",
       "\n",
       "   experiment_id              title                  organism  assay_id  \\\n",
       "0      SRX717562        input2.2_15  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "1      SRX717561        input1.2_15  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "2      SRX717560         input2.2_8  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "3      SRX717559         input3.1_8  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "4      SRX717558        input1.3_60  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "..           ...                ...                       ...       ...   \n",
       "1      SRX127920  delta_chd1_K79me3  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "2      SRX127919     delta_chd1_wce  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "3      SRX127918     wt_FM391_K4me3  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "4      SRX127917    wt_FM391_K79me3  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "5      SRX127916       wt_FM391_wce  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "\n",
       "                                           attributes  \n",
       "0   source_name : BY4741 stress : diamide time : 1...  \n",
       "1   source_name : BY4741 stress : diamide time : 1...  \n",
       "2   source_name : BY4741 stress : diamide time : 8...  \n",
       "3   source_name : BY4741 stress : diamide time : 8...  \n",
       "4   source_name : BY4741 stress : diamide time : 6...  \n",
       "..                                                ...  \n",
       "1   source_name : delta_Chd1 yeast (FM391,delta Ch...  \n",
       "2   source_name : delta_Chd1 yeast (FM391,delta Ch...  \n",
       "3   source_name : Wild-type yeast (FM391) genotype...  \n",
       "4   source_name : Wild-type yeast (FM391) genotype...  \n",
       "5   source_name : Wild-type yeast (FM391) genotype...  \n",
       "\n",
       "[1172 rows x 10 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = sragent.gather(projects)\n",
    "#meta = pd.read_csv('sragent_output/metadata.csv')\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`meta` is a pandas dataframe with all of the project ids, experiment ids, abstracts, protocols, experiment titles, and experiment attributes.\\\n",
    "There are a total of 1,172 experiments here.\\\n",
    "\\\n",
    "`annotate()` can also take this dataframe as input, that way we don't have to unnecessarily repeat metadata pulls from the SRA.\\\n",
    "It also let's us manipulate an subset the metadata before we annotate. \\\n",
    "For instance, in this vignette we have a total of 25 different projects that include histone PTM ChIP-seq experiments in yeast. \\\n",
    "But there are non-histone ChIP-seq experiments within these projects as well that we don't care about for now. \\\n",
    "Let's try to filter down to just experiments with histone targets.\\"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First we'll read in a csv with all the histone ptm targets we're interested in, as well as 'H3' and 'input' targets so we don't lose our ChIP-seq controls.\\\n",
    "We then filter the metadata dataframe with a single string pattern of our desired epitopes and using that as a mask. \\\n",
    "That drops our total number of experiments to annotate to 899. \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>project_id</th>\n",
       "      <th>project_title</th>\n",
       "      <th>abstract</th>\n",
       "      <th>protocol</th>\n",
       "      <th>run_id</th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>title</th>\n",
       "      <th>organism</th>\n",
       "      <th>assay_id</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593257</td>\n",
       "      <td>SRX717562</td>\n",
       "      <td>input2.2_15</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593256</td>\n",
       "      <td>SRX717561</td>\n",
       "      <td>input1.2_15</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593255</td>\n",
       "      <td>SRX717560</td>\n",
       "      <td>input2.2_8</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593254</td>\n",
       "      <td>SRX717559</td>\n",
       "      <td>input3.1_8</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593253</td>\n",
       "      <td>SRX717558</td>\n",
       "      <td>input1.3_60</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 6...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>190</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593061</td>\n",
       "      <td>SRX717366</td>\n",
       "      <td>h2ak5ac_4</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 4...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>191</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593060</td>\n",
       "      <td>SRX717365</td>\n",
       "      <td>h2ak5ac_8</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 8...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593059</td>\n",
       "      <td>SRX717364</td>\n",
       "      <td>h2ak5ac_15</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593058</td>\n",
       "      <td>SRX717363</td>\n",
       "      <td>h2ak5ac_30</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 3...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194</th>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>High resolution chromatin dynamics during a ye...</td>\n",
       "      <td>Covalent histone modifications are highly cons...</td>\n",
       "      <td>Formaldehyde was quenched by 125mM glycine. Ce...</td>\n",
       "      <td>SRR1593057</td>\n",
       "      <td>SRX717362</td>\n",
       "      <td>h2ak5ac_60</td>\n",
       "      <td>Saccharomyces_cerevisiae</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>source_name : BY4741 stress : diamide time : 6...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>195 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      project_id                                      project_title  \\\n",
       "0    PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "1    PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "2    PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "3    PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "4    PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "..           ...                                                ...   \n",
       "190  PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "191  PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "192  PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "193  PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "194  PRJNA262623  High resolution chromatin dynamics during a ye...   \n",
       "\n",
       "                                              abstract  \\\n",
       "0    Covalent histone modifications are highly cons...   \n",
       "1    Covalent histone modifications are highly cons...   \n",
       "2    Covalent histone modifications are highly cons...   \n",
       "3    Covalent histone modifications are highly cons...   \n",
       "4    Covalent histone modifications are highly cons...   \n",
       "..                                                 ...   \n",
       "190  Covalent histone modifications are highly cons...   \n",
       "191  Covalent histone modifications are highly cons...   \n",
       "192  Covalent histone modifications are highly cons...   \n",
       "193  Covalent histone modifications are highly cons...   \n",
       "194  Covalent histone modifications are highly cons...   \n",
       "\n",
       "                                              protocol      run_id  \\\n",
       "0    Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593257   \n",
       "1    Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593256   \n",
       "2    Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593255   \n",
       "3    Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593254   \n",
       "4    Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593253   \n",
       "..                                                 ...         ...   \n",
       "190  Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593061   \n",
       "191  Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593060   \n",
       "192  Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593059   \n",
       "193  Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593058   \n",
       "194  Formaldehyde was quenched by 125mM glycine. Ce...  SRR1593057   \n",
       "\n",
       "    experiment_id        title                  organism  assay_id  \\\n",
       "0       SRX717562  input2.2_15  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "1       SRX717561  input1.2_15  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "2       SRX717560   input2.2_8  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "3       SRX717559   input3.1_8  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "4       SRX717558  input1.3_60  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "..            ...          ...                       ...       ...   \n",
       "190     SRX717366    h2ak5ac_4  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "191     SRX717365    h2ak5ac_8  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "192     SRX717364   h2ak5ac_15  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "193     SRX717363   h2ak5ac_30  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "194     SRX717362   h2ak5ac_60  Saccharomyces_cerevisiae  ChIP-Seq   \n",
       "\n",
       "                                            attributes  \n",
       "0    source_name : BY4741 stress : diamide time : 1...  \n",
       "1    source_name : BY4741 stress : diamide time : 1...  \n",
       "2    source_name : BY4741 stress : diamide time : 8...  \n",
       "3    source_name : BY4741 stress : diamide time : 8...  \n",
       "4    source_name : BY4741 stress : diamide time : 6...  \n",
       "..                                                 ...  \n",
       "190  source_name : BY4741 stress : diamide time : 4...  \n",
       "191  source_name : BY4741 stress : diamide time : 8...  \n",
       "192  source_name : BY4741 stress : diamide time : 1...  \n",
       "193  source_name : BY4741 stress : diamide time : 3...  \n",
       "194  source_name : BY4741 stress : diamide time : 6...  \n",
       "\n",
       "[195 rows x 10 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "epitopes = pd.read_csv('epitopes.csv')\n",
    "eps = '|'.join(epitopes['epitope_id'].tolist())\n",
    "mask = meta['title'].str.contains(eps, case=False, na=False)\n",
    "ptm_meta = meta[mask]\n",
    "ptm_meta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When we're happy with our subset of experiments we can then rerun `annotate()` with the subset metadata dataframe as input.\\\n",
    "By default `annotate()` will save the annotation as a csv, but we can store the resulting dataframe to explore the results. \\"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Let's break down the metadata and identify the key elements that will help classify and generate structured metadata for each experiment in the project.\n",
      "\n",
      "### Main Goal of the Project\n",
      "The main goal of the project is to map 26 histone modifications genome-wide over a time course following dramatic transcriptional reprogramming in response to diamide stress in yeast. The project aims to understand the dynamics and combinatorial complexity of histone modifications during the stress response.\n",
      "\n",
      "### Experimental Conditions\n",
      "- **Stress Condition**: Diamide stress\n",
      "- **Time Points**: 0, 4, 8, 15, 30, 60 minutes\n",
      "- **Controls**: MNase input controls at each time point\n",
      "- **Histone Modifications**: 26 different histone modifications\n",
      "\n",
      "### Key Words and Their Indications\n",
      "\n",
      "#### Gene Mutations\n",
      "- **Key Words**: There are no specific keywords indicating gene mutations in the provided metadata. The strain used is BY4741, which is a common yeast strain and does not indicate any specific mutations relevant to the project goal.\n",
      "\n",
      "#### Gene Deletions\n",
      "- **Key Words**: There are no specific keywords indicating gene deletions in the provided metadata. The strain used is BY4741, which does not indicate any specific gene deletions relevant to the project goal.\n",
      "\n",
      "#### Protein Depletions\n",
      "- **Key Words**: There are no specific keywords indicating protein depletions in the provided metadata. The project does not mention any protein depletion experiments.\n",
      "\n",
      "#### Chemical Treatments or Stress Conditions\n",
      "- **Key Words**: \"diamide\" indicates the chemical treatment used to induce stress in the yeast cells. This signifies that the cells were subjected to oxidative stress.\n",
      "\n",
      "#### Time Points or Stages of Development/Growth\n",
      "- **Key Words**: \"time\" followed by a specific minute value (e.g., \"0 min\", \"4 min\", \"8 min\", \"15 min\", \"30 min\", \"60 min\") indicates the different time points at which samples were collected during the stress response.\n",
      "\n",
      "#### ChIP Input Control\n",
      "- **Key Words**: \"input\" in the experiment title indicates that the experiment is a ChIP Input control. For example, \"input1.1_0\" or \"input2.2_15\".\n",
      "\n",
      "#### Non-Specific Antibody Control\n",
      "- **Key Words**: There are no specific keywords indicating non-specific antibody controls in the provided metadata.\n",
      "\n",
      "#### ChIP-seq Protein Targets\n",
      "- **Key Words**: The experiment titles and attributes contain the specific histone modifications being targeted. For example:\n",
      "  - \"h4r3me2s\" indicates the histone modification H4R3me2s.\n",
      "  - \"h4k8ac\" indicates the histone modification H4K8ac.\n",
      "  - \"h3k4me3\" indicates the histone modification H3K4me3.\n",
      "  - If an experiment is identified as 'Input', the protein target must be 'None'.\n",
      "\n",
      "#### Experimental Replicates\n",
      "- **Key Words**: \"experimental_batch\" followed by a specific value (e.g., \"1.1\", \"2.2\", \"3.1\") indicates different experimental replicates.\n",
      "\n",
      "#### Small Molecules Used\n",
      "- **Key Words**: \"diamide\" is the small molecule used to induce stress in the yeast cells.\n",
      "\n",
      "### Structured Metadata Elements\n",
      "Based on the identified key words, the structured metadata for each experiment should include the following elements:\n",
      "- **Experiment ID**\n",
      "- **Experiment Title**\n",
      "- **Source Name**: The yeast strain used (e.g., BY4741)\n",
      "- **Stress Condition**: The chemical treatment used (e.g., diamide)\n",
      "- **Time Point**: The time at which the sample was collected (e.g., 0 min, 4 min, 8 min, 15 min, 30 min, 60 min)\n",
      "- **Antibody**: The antibody used for ChIP-seq (e.g., anti-h4r3me2s, anti-h4k8ac)\n",
      "- **Experimental Batch**: The replicate batch (e.g., 1.1, 2.2, 3.1)\n",
      "- **ChIP-seq Protein Target**: The specific histone modification or 'None' for input controls (e.g., H4R3me2s, H4K8ac, None)\n",
      "- **Control Type**: Indicate if the experiment is an input control (e.g., Input Control, None)\n",
      "\n",
      "### Example Structured Metadata for an Experiment\n",
      "Here is an example of structured metadata for one of the experiments:\n",
      "\n",
      "**Experiment ID**: SRX717562\n",
      "- **Experiment Title**: input2.2_15\n",
      "- **Source Name**: BY4741\n",
      "- **Stress Condition**: diamide\n",
      "- **Time Point**: 15 min\n",
      "- **Antibody**: na\n",
      "- **Experimental Batch**: 2.2\n",
      "- **ChIP-seq Protein Target**: None\n",
      "- **Control Type**: Input Control\n",
      "\n",
      "By following this approach, you can classify and generate structured metadata for each experiment in the project.\n",
      "annotating experiment 0    SRX717562\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 1    SRX717561\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 2    SRX717560\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 3    SRX717559\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 4    SRX717558\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 5    SRX717557\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 6    SRX717556\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 7    SRX717555\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 8    SRX717554\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 9    SRX717553\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 10    SRX717552\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 11    SRX717551\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 12    SRX717550\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 13    SRX717549\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 14    SRX717548\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 15    SRX717547\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 16    SRX717546\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 17    SRX717545\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 18    SRX717544\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 19    SRX717543\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 20    SRX717542\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 21    SRX717541\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 22    SRX717540\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 23    SRX717539\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 24    SRX717538\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 25    SRX717537\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 26    SRX717536\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 27    SRX717535\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 28    SRX717534\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 29    SRX717533\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 30    SRX717532\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 31    SRX717531\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 32    SRX717530\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 33    SRX717529\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 34    SRX717528\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 35    SRX717527\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 36    SRX717526\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 37    SRX717525\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 38    SRX717524\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 39    SRX717523\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 40    SRX717522\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 41    SRX717521\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 42    SRX717520\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 43    SRX717519\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 44    SRX717518\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 45    SRX717511\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 46    SRX717510\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 47    SRX717509\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 48    SRX717508\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 49    SRX717507\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 50    SRX717506\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 51    SRX717505\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 52    SRX717504\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 53    SRX717503\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 54    SRX717502\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 55    SRX717501\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 56    SRX717500\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 57    SRX717499\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 58    SRX717498\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 59    SRX717497\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 60    SRX717496\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 61    SRX717495\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 62    SRX717494\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 63    SRX717493\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 64    SRX717492\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 65    SRX717491\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 66    SRX717490\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 67    SRX717489\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 68    SRX717488\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 69    SRX717487\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 70    SRX717486\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 71    SRX717485\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 72    SRX717484\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 73    SRX717483\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 74    SRX717482\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 75    SRX717481\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 76    SRX717480\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 77    SRX717479\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 78    SRX717478\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 79    SRX717477\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 80    SRX717476\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 81    SRX717475\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 82    SRX717474\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 83    SRX717473\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 84    SRX717472\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 85    SRX717471\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 86    SRX717470\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 87    SRX717469\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 88    SRX717468\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 89    SRX717467\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 90    SRX717466\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 91    SRX717465\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 92    SRX717464\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 93    SRX717463\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 94    SRX717462\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 95    SRX717461\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 96    SRX717460\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 97    SRX717459\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 98    SRX717458\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 99    SRX717457\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 100    SRX717456\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 101    SRX717455\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 102    SRX717454\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 103    SRX717453\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 104    SRX717452\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 105    SRX717451\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 106    SRX717450\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 107    SRX717449\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 108    SRX717448\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 109    SRX717447\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 110    SRX717446\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 111    SRX717445\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 112    SRX717444\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 113    SRX717443\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 114    SRX717442\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 115    SRX717441\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 116    SRX717440\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 117    SRX717439\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 118    SRX717438\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 119    SRX717437\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 120    SRX717436\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 121    SRX717435\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 122    SRX717434\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 123    SRX717433\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 124    SRX717432\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 125    SRX717431\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 126    SRX717430\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 127    SRX717429\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 128    SRX717428\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 129    SRX717427\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 130    SRX717426\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 131    SRX717425\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 132    SRX717424\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 133    SRX717423\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 134    SRX717422\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 135    SRX717421\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 136    SRX717420\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 137    SRX717419\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 138    SRX717418\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 139    SRX717417\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 140    SRX717416\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 141    SRX717415\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 142    SRX717414\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 143    SRX717413\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 144    SRX717412\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 145    SRX717411\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 146    SRX717410\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 147    SRX717409\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 148    SRX717408\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 149    SRX717407\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 150    SRX717406\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 151    SRX717405\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 152    SRX717404\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 153    SRX717403\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 154    SRX717402\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 155    SRX717401\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 156    SRX717400\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 157    SRX717399\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 158    SRX717398\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 159    SRX717397\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 160    SRX717396\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 161    SRX717395\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 162    SRX717394\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 163    SRX717393\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 164    SRX717392\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 165    SRX717391\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 166    SRX717390\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 167    SRX717389\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 168    SRX717388\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 169    SRX717387\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 170    SRX717386\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 171    SRX717385\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 172    SRX717384\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 173    SRX717383\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 174    SRX717382\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 175    SRX717381\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 176    SRX717380\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 177    SRX717379\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 178    SRX717378\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 179    SRX717377\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 180    SRX717376\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 181    SRX717375\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 182    SRX717374\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 183    SRX717373\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 184    SRX717372\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 185    SRX717371\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 186    SRX717370\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 187    SRX717369\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 188    SRX717368\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 189    SRX717367\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 190    SRX717366\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 191    SRX717365\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 192    SRX717364\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 193    SRX717363\n",
      "Name: experiment_id, dtype: object...\n",
      "annotating experiment 194    SRX717362\n",
      "Name: experiment_id, dtype: object...\n",
      "project PRJNA262623 done!\n",
      "tagging...\n",
      "setting controls...\n"
     ]
    }
   ],
   "source": [
    "#test = ptm_meta[ptm_meta['project_id'] == 'PRJNA262623']\n",
    "ptm_meta_annotated = sragent.annotate(ptm_meta, \n",
    "                                      model_summary = 'gpt-4o-mini',\n",
    "                                      model_annotation = 'gpt-4o-2024-08-06',\n",
    "                                      annotate = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this large example, using `gpt-4o-mini` as of 08/13/24, annotation took 33m54s, using ~1.8M tokens (1.68M context, 0.1M generated) at a total cost of ~$0.30. \\\n",
    "For the larger `gpt-4o` model (is currently ~35x more expensive) we can ballpark the same annotation would cost ~$10.5 \\ \n",
    "This is a rough estimate because input and output cost differs substantially, but gives a decent idea. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So what did `annotate()` do?\n",
    "Let's take a look at the output, which should be saved as `annotation_FULL.csv` in the `sragent_output` directory."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`ptm_meta_annotated` is a pandas dataframe with the following columns: \\\n",
    "- `project_id`\n",
    "- `experiment_id`\n",
    "- `exp_title` - title of the experiment as listed in the SRA metadata\n",
    "(the following columns are generated from the LLM response)\n",
    "- `gene_mutatation` - True or False: does this experiment test a gene mutation? \n",
    "- `gene_deletion` - True or False: does this experiment test a gene deletion?\n",
    "- `protein_depletion` - True or False: does this experiment test a protein depletion?\n",
    "- `stress_condition` - True or False: does this experiment test a stress condition?\n",
    "- `time_series` - True or False: is this experiment specify a specific time point or growth stage?\n",
    "- `chip_input` - True or False: is this an input experiment?\n",
    "- `antibody_control` - True or False: is this an antibody control? ie. a non-specific antibody\n",
    "- `chip_target` - protein targeted in the experiment\n",
    "- `mutatation` - specific mutation if present\n",
    "- `deletion` - specific deletion if present\n",
    "- `depletion` - protein depleted if present\n",
    "- `stress` - stress condition if present\n",
    "- `time_point`\n",
    "(the following are determined based on the classifications from the LLM above)\n",
    "- `warnning` - True or False: flags a logical disagreement between the LLM classifications. Useful to catch mistakes and guide human review\n",
    "- `sample` - a simple sample name or tag that merges experiment attributes like genotype, time, target. \n",
    "- `perturbation` - one of ['None','gene_mutation','gene_deletion','protein_depletion','stress_condition']\n",
    "- `control` - the `sample` name of the control (input or other) experiment that matches target for that project\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's examine how well the `gpt-4o-mini` model did at classifying and annotating these experiments.\\\n",
    "How many experiments were flagged for review?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>experiment_id</th>\n",
       "      <th>exp_title</th>\n",
       "      <th>gene_mutation</th>\n",
       "      <th>gene_deletion</th>\n",
       "      <th>protein_depletion</th>\n",
       "      <th>stress_condition</th>\n",
       "      <th>time_series</th>\n",
       "      <th>chip_input</th>\n",
       "      <th>antibody_control</th>\n",
       "      <th>chip_target</th>\n",
       "      <th>...</th>\n",
       "      <th>deletion</th>\n",
       "      <th>depletion</th>\n",
       "      <th>stress</th>\n",
       "      <th>time_point</th>\n",
       "      <th>project_id</th>\n",
       "      <th>model</th>\n",
       "      <th>warning</th>\n",
       "      <th>sample</th>\n",
       "      <th>perturbation</th>\n",
       "      <th>control</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>project_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">PRJNA262623</th>\n",
       "      <th>0</th>\n",
       "      <td>SRX717474</td>\n",
       "      <td>h4k12ac_4</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>H4K12ac</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>diamide</td>\n",
       "      <td>4 min</td>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>False</td>\n",
       "      <td>H4K12ac-diamide-stress_condition-4_min</td>\n",
       "      <td>stress_condition</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRX717446</td>\n",
       "      <td>h3k79me_60</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>H3K79me</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>diamide</td>\n",
       "      <td>60 min</td>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>False</td>\n",
       "      <td>H3K79me-diamide-stress_condition-60_min</td>\n",
       "      <td>stress_condition</td>\n",
       "      <td>Input-diamide-stress_condition-60_min</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRX717506</td>\n",
       "      <td>h4r3me2s_60</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>H4R3me2s</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>diamide</td>\n",
       "      <td>60 min</td>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>False</td>\n",
       "      <td>H4R3me2s-diamide-stress_condition-60_min</td>\n",
       "      <td>stress_condition</td>\n",
       "      <td>Input-diamide-stress_condition-60_min</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRX717485</td>\n",
       "      <td>h4k20me_8</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>H4K20me</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>diamide</td>\n",
       "      <td>8 min</td>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>False</td>\n",
       "      <td>H4K20me-diamide-stress_condition-8_min</td>\n",
       "      <td>stress_condition</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRX717537</td>\n",
       "      <td>input1.2_60</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>diamide</td>\n",
       "      <td>60 min</td>\n",
       "      <td>PRJNA262623</td>\n",
       "      <td>gpt-4o</td>\n",
       "      <td>False</td>\n",
       "      <td>Input-diamide-stress_condition-60_min</td>\n",
       "      <td>stress_condition</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              experiment_id    exp_title  gene_mutation  gene_deletion  \\\n",
       "project_id                                                               \n",
       "PRJNA262623 0     SRX717474    h4k12ac_4          False          False   \n",
       "            2     SRX717446   h3k79me_60          False          False   \n",
       "            3     SRX717506  h4r3me2s_60          False          False   \n",
       "            4     SRX717485    h4k20me_8          False          False   \n",
       "            1     SRX717537  input1.2_60          False          False   \n",
       "\n",
       "               protein_depletion  stress_condition  time_series  chip_input  \\\n",
       "project_id                                                                    \n",
       "PRJNA262623 0              False              True         True       False   \n",
       "            2              False              True         True       False   \n",
       "            3              False              True         True       False   \n",
       "            4              False              True         True       False   \n",
       "            1              False              True         True        True   \n",
       "\n",
       "               antibody_control chip_target  ... deletion depletion   stress  \\\n",
       "project_id                                   ...                               \n",
       "PRJNA262623 0             False     H4K12ac  ...                     diamide   \n",
       "            2             False     H3K79me  ...                     diamide   \n",
       "            3             False    H4R3me2s  ...                     diamide   \n",
       "            4             False     H4K20me  ...                     diamide   \n",
       "            1             False              ...                     diamide   \n",
       "\n",
       "              time_point   project_id   model warning  \\\n",
       "project_id                                              \n",
       "PRJNA262623 0      4 min  PRJNA262623  gpt-4o   False   \n",
       "            2     60 min  PRJNA262623  gpt-4o   False   \n",
       "            3     60 min  PRJNA262623  gpt-4o   False   \n",
       "            4      8 min  PRJNA262623  gpt-4o   False   \n",
       "            1     60 min  PRJNA262623  gpt-4o   False   \n",
       "\n",
       "                                                 sample      perturbation  \\\n",
       "project_id                                                                  \n",
       "PRJNA262623 0    H4K12ac-diamide-stress_condition-4_min  stress_condition   \n",
       "            2   H3K79me-diamide-stress_condition-60_min  stress_condition   \n",
       "            3  H4R3me2s-diamide-stress_condition-60_min  stress_condition   \n",
       "            4    H4K20me-diamide-stress_condition-8_min  stress_condition   \n",
       "            1     Input-diamide-stress_condition-60_min  stress_condition   \n",
       "\n",
       "                                             control  \n",
       "project_id                                            \n",
       "PRJNA262623 0                                   None  \n",
       "            2  Input-diamide-stress_condition-60_min  \n",
       "            3  Input-diamide-stress_condition-60_min  \n",
       "            4                                   None  \n",
       "            1                                    NaN  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ptm_meta_annotated\n",
    "#print(ptm_meta_annotated[ptm_meta_annotated['warning']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sragent",
   "language": "python",
   "name": "sragent"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
